src/gt4py/backend/dawn_backends.py

```diff
@@ -196,24 +251,113 @@ def visit_StencilDefinition(self, node: gt_ir.StencilDefinition, **kwargs):
         return sir
 
 
-convert_to_SIR = SIRConverter.apply
+_DAWN_BASE_OPTIONS = {
+    "add_profile_info": {"versioning": True},
+    "clean": {"versioning": False},
+    "debug_mode": {"versioning": True},
+    "dump_sir": {"versioning": False},
+    "verbose": {"versioning": False},
+    "no_opt": {"versioning": False},
+}
+
+
+_DAWN_TOOLCHAIN_OPTIONS = {}
+for name in dir(dawn4py.CodeGenOptions) + dir(dawn4py.OptimizerOptions):
+    if (
+        name.startswith("print")
+        or name.startswith("dump")
+        or name.startswith("report")
+        or name.startswith("serialize")
+        or name.startswith("deserialize")
+    ):
+        _DAWN_TOOLCHAIN_OPTIONS[name] = {"versioning": False}
+    elif not name.startswith("_") and name != "backend":
+        _DAWN_TOOLCHAIN_OPTIONS[name] = {"versioning": True}
+
+
+_DAWN_BACKEND_OPTIONS = {**_DAWN_BASE_OPTIONS, **_DAWN_TOOLCHAIN_OPTIONS}
+
+
+class DawnPyModuleGenerator(gt_backend.PyExtModuleGenerator):
+    def __init__(self, backend_class):
+        super().__init__(backend_class)
+
+    def generate_implementation(self):
+        sources = gt_text.TextBlock(
+            indent_size=gt_backend.BaseModuleGenerator.TEMPLATE_INDENT_SIZE
+        )
+
+        args = []
+        empty_checks = []
+        none_checks = []
+
+        for arg in self.implementation_ir.api_signature:
+            args.append(arg.name)
+            if arg.name in self.implementation_ir.fields:
+                if arg.name in self.implementation_ir.unreferenced:
+                    ndims = len(self.implementation_ir.fields[arg.name].axes)
+                    zeros = ", ".join(["0"] * ndims)
+                    args.append("[{}]".format(zeros))
+                    empty_checks.append(
+                        f"{arg.name} = np.empty(({zeros})) if {arg.name} is None else {arg.name}"
+                    )
+                else:
+                    args.append("list(_origin_['{}'])".format(arg.name))
+            elif arg.name in self.implementation_ir.parameters and arg.default == None:
+                none_checks.append(f"{arg.name} = 0 if {arg.name} is None else {arg.name}")
+
+        source = """
+# Load or generate a Dawn Computation object for the current domain size
+{empty_checks}
+{none_checks}
+pyext_module.run_computation(list(_domain_), {run_args}, exec_info)
+""".format(
+            empty_checks="\n".join(empty_checks),
+            none_checks="\n".join(none_checks),
+            run_args=", ".join(args),
+        )
+        sources.extend(source.splitlines())
```

Comment:
I guess this code is creating default values for the case where optional parameters/fields have not been provided by the user, right? Could you add some comment here to make it clear what this code does and why ?

Reply:
Comments have been updated.